src/main/scala/vexiiriscv/execute/IterativeShiftPlugin.scala

```diff
@@ -0,0 +1,201 @@
+        layer(op).addDecoding(IS_W -> False, IS_W_RIGHT -> False)
+      }
+      add(Rvi.SLLW).srcs(SRC1.RF, SRC2.RF).decode(LEFT -> True, ARITHMETIC -> False, IS_W -> True, IS_W_RIGHT -> False)
+      add(Rvi.SRLW).srcs(SRC1.RF, SRC2.RF).decode(LEFT -> False, ARITHMETIC -> False, IS_W -> True, IS_W_RIGHT -> True)
+      add(Rvi.SRAW).srcs(SRC1.RF, SRC2.RF).decode(LEFT -> False, ARITHMETIC -> True, IS_W -> True, IS_W_RIGHT -> True)
+      add(Rvi.SLLIW).srcs(SRC1.RF, SRC2.I).decode(LEFT -> True, ARITHMETIC -> False, IS_W -> True, IS_W_RIGHT -> False)
+      add(Rvi.SRLIW).srcs(SRC1.RF, SRC2.I).decode(LEFT -> False, ARITHMETIC -> False, IS_W -> True, IS_W_RIGHT -> True)
+      add(Rvi.SRAIW).srcs(SRC1.RF, SRC2.I).decode(LEFT -> False, ARITHMETIC -> True, IS_W -> True, IS_W_RIGHT -> True)
+      for (op <- List(Rvi.SLLW, Rvi.SRLW, Rvi.SRAW, Rvi.SLLIW, Rvi.SRLIW, Rvi.SRAIW)) {
+        ifp.signExtend(wb, layer(op), 32)
+      }
+    }
+
+    uopRetainer.release()
+
+    val shift = new el.Execute(shiftAt) {
+      val unscheduleRequest = RegNext(isCancel).clearWhen(isReady).init(False)
+      val selected = isValid && SEL
+
+      val amplitudeWidth = if(Riscv.XLEN.get == 64) 6 else 5
+      val shamt = if(Riscv.XLEN.get==32) {
+        srcp.SRC2.resize(amplitudeWidth bit).asUInt
+      } else {
+        // SxxW instructions only use lower 5 bit of SRC2, not 6
+        (srcp.SRC2.resize(amplitudeWidth bit).asBits & B(6 bit, 5 -> !IS_W, default -> True)).asUInt
+      }
+      val rs1 = el(IntRegFile, RS1).asBits
+
+      val busy = RegInit(False)
+      val flipped = Reg(Bool())
+      val amplitude = Reg(cloneOf(shamt))
+      val shiftReg = Reg(cloneOf(srcp.SRC1.asBits))
+
+      val done = if(lateResult) {
+        amplitude === 0 & !flipped & busy
+      } else {
+        val rightShiftDone = !LEFT & busy & rightShifts.map(amplitude === _).reduce(_ | _)
+        val leftShiftDone = if(leftShifts.isEmpty) {
+          // we need to fully shift since we still need to flip back
+          busy & amplitude === 0
+        } else {
+          busy & leftShifts.map(amplitude === _).reduce(_ | _)
+        }
+
+        // done comes one cycle "early", in the cycle we do the last action (shift / flip)
+        shamt === 0 | leftShiftDone | rightShiftDone
+      }
+
+      val signExtended =
+        if(Riscv.XLEN.get == 64)
+          IS_W_RIGHT ? (B(32 bit, default -> (ARITHMETIC & rs1(31))) ## rs1(31 downto 0)) | rs1
+        else
+          CombInit(rs1)
+
+      val muxInputs = mutable.ArrayBuffer[(Bool, Bits, UInt)]()
+      rightShifts.sorted.foreach(n => {
+        val doIt = if(n > 1) amplitude >= n else True
+        val input = (((ARITHMETIC && srcp.SRC1.msb) #* n) ## shiftReg) >> n
+        val ampl = amplitude - n
+        muxInputs.append((doIt, input, ampl))
+      })
+
+      leftShifts.sorted.foreach(n => {
+        // mask away bits shifted into upper 32 bit in case of SLLW instruction
```

Comment:
Is it necessary ? As anyway, the `ifp.signExtend(wb, layer(op), 32)` will erase all upper bits using the bit 31 as bit sign.

Reply:
fixed